### PR TITLE
mips: fix regression after "Handle Gstep according the N64/N32 ABI"

### PR DESCRIPTION
**Pull request by ksergeyv**
unw_step should return zero for _MIPS_SIM != _ABI64 when dwarf_step failed.
This restores unw_step behaviour before 5eec9a2ecb9a93996d566bbfbcdbe006f64b7e16
for _ABIN32 and _ABIO32 interfaces.

### CODE REVIEW

**Comment by kasperk81** [comment]
@patrickdepinguin, does this change fix #173 and #172 for you?

**Comment by ksergeyv** [comment]
@patrickdepinguin @kasperk81 please note that this patch may not be sufficient for mips at least.
In my mips/mipsel (O32) and aarch64 environments with gcc 10.3 revert of #220 required also.

**Comment by kasperk81** [comment]
@ksergeyv, do you have a sample where aarch64 is broken without #220 revert? Maybe we can add it as a test in a separate PR.

**Comment by patrickdepinguin** [comment]
@kasperk81 This patch does improve the situation for remote unwinding (issue #173 ) but does not make a difference for the endless last entry of local unwinding (oddly enough).

Below is the output I see in different cases for two test programs:

* backtrace_test_libunwind_ptrace: remote unwinding, as described in issue #173 .
* backtrace_test_libunwind: local unwinding, as described in issue #172 .

1. libunwind 1.4.0
```
$ LD_LIBRARY_PATH=/tmp/unwind  /tmp/unwind/backtrace_test_libunwind_ptrace $(pidof dropbear) | head -n 20
pid: 21898
0000000076fe4ce4 __select                         offset 0x44 (sp=000000007fe47a50)
000000001000c8d4 _init                            offset 0x9964 (sp=000000007fe47aa0)
000000001001f024 _init                            offset 0x1c0b4 (sp=000000007fe47db0)
000000001001e240 _init                            offset 0x1b2d0 (sp=000000007fe47df0)
000000001001e2fc _init                            offset 0x1b38c (sp=000000007fe480f0)
0000000076f13e1c __libc_start_main                offset 0xfc (sp=000000007fe48120)
0000000010003c4c _init                            offset 0xcdc (sp=000000007fe48220)
0000000010003c4c _init                            offset 0xcdc (sp=000000007fe48220)
0000000010003c4c _init                            offset 0xcdc (sp=000000007fe48220)
0000000010003c4c _init                            offset 0xcdc (sp=000000007fe48220)
0000000010003c4c _init                            offset 0xcdc (sp=000000007fe48220)
0000000010003c4c _init                            offset 0xcdc (sp=000000007fe48220)
0000000010003c4c _init                            offset 0xcdc (sp=000000007fe48220)
0000000010003c4c _init                            offset 0xcdc (sp=000000007fe48220)
0000000010003c4c _init                            offset 0xcdc (sp=000000007fe48220)
0000000010003c4c _init                            offset 0xcdc (sp=000000007fe48220)
0000000010003c4c _init                            offset 0xcdc (sp=000000007fe48220)
0000000010003c4c _init                            offset 0xcdc (sp=000000007fe48220)
0000000010003c4c _init                            offset 0xcdc (sp=000000007fe48220)

$ LD_LIBRARY_PATH=/tmp/unwind  /tmp/unwind/backtrace_test_libunwind | head -n 20
#1  0x0000000010001218 sp=0x000000007fff17a0 fac(int) + 0x28
#2  0x0000000010001238 sp=0x000000007fff17c0 fac(int) + 0x48
#3  0x0000000010001238 sp=0x000000007fff17e0 fac(int) + 0x48
#4  0x0000000010001238 sp=0x000000007fff1800 fac(int) + 0x48
#5  0x0000000010001238 sp=0x000000007fff1820 fac(int) + 0x48
#6  0x0000000010001238 sp=0x000000007fff1840 fac(int) + 0x48
#7  0x0000000010001238 sp=0x000000007fff1860 fac(int) + 0x48
#8  0x0000000010001238 sp=0x000000007fff1880 fac(int) + 0x48
#9  0x0000000010001238 sp=0x000000007fff18a0 fac(int) + 0x48
#10 0x0000000010001238 sp=0x000000007fff18c0 fac(int) + 0x48
#11 0x0000000010001238 sp=0x000000007fff18e0 fac(int) + 0x48
#12 0x00000000100013d0 sp=0x000000007fff1900 main + 0x174
#13 0x00000000777ede1c sp=0x000000007fff1990 __libc_start_main + 0xfc
#14 0x0000000010000f5c sp=0x000000007fff1a90 __start + 0x4c
#15 0x0000000010000f5c sp=0x000000007fff1a90 __start + 0x4c
#16 0x0000000010000f5c sp=0x000000007fff1a90 __start + 0x4c
#17 0x0000000010000f5c sp=0x000000007fff1a90 __start + 0x4c
#18 0x0000000010000f5c sp=0x000000007fff1a90 __start + 0x4c
#19 0x0000000010000f5c sp=0x000000007fff1a90 __start + 0x4c
#20 0x0000000010000f5c sp=0x000000007fff1a90 __start + 0x4c
```
In both cases we suffer from the endless last entry.

2. When reverting commit 5eec9a2ecb9a93996d566bbfbcdbe006f64b7e16 on top of libunwind-1.4.0 :
```
$ LD_LIBRARY_PATH=/tmp/unwind  /tmp/unwind/backtrace_test_libunwind_ptrace $(pidof dropbear) | head -n 20
pid: 21898
0000000076fe4ce4 __select                         offset 0x44 (sp=000000007fe47a50)
000000001000c8d4 _init                            offset 0x9964 (sp=000000007fe47aa0)
000000001001f024 _init                            offset 0x1c0b4 (sp=000000007fe47db0)
000000001001e240 _init                            offset 0x1b2d0 (sp=000000007fe47df0)
000000001001e2fc _init                            offset 0x1b38c (sp=000000007fe480f0)
0000000076f13e1c __libc_start_main                offset 0xfc (sp=000000007fe48120)
0000000010003c4c _init                            offset 0xcdc (sp=000000007fe48220)

$ LD_LIBRARY_PATH=/tmp/unwind  /tmp/unwind/backtrace_test_libunwind | head -n 20
#1  0x0000000010001218 sp=0x000000007ffe42e0 fac(int) + 0x28
#2  0x0000000010001238 sp=0x000000007ffe4300 fac(int) + 0x48
#3  0x0000000010001238 sp=0x000000007ffe4320 fac(int) + 0x48
#4  0x0000000010001238 sp=0x000000007ffe4340 fac(int) + 0x48
#5  0x0000000010001238 sp=0x000000007ffe4360 fac(int) + 0x48
#6  0x0000000010001238 sp=0x000000007ffe4380 fac(int) + 0x48
#7  0x0000000010001238 sp=0x000000007ffe43a0 fac(int) + 0x48
#8  0x0000000010001238 sp=0x000000007ffe43c0 fac(int) + 0x48
#9  0x0000000010001238 sp=0x000000007ffe43e0 fac(int) + 0x48
#10 0x0000000010001238 sp=0x000000007ffe4400 fac(int) + 0x48
#11 0x0000000010001238 sp=0x000000007ffe4420 fac(int) + 0x48
#12 0x00000000100013d0 sp=0x000000007ffe4440 main + 0x174
#13 0x000000007789ee1c sp=0x000000007ffe44d0 __libc_start_main + 0xfc
#14 0x0000000010000f5c sp=0x000000007ffe45d0 __start + 0x4c
```

This shows the expected behavior in both cases.

3. With the patch proposed by @ksergeyv on top of libunwind-1.4.0:
```
$ LD_LIBRARY_PATH=/tmp/unwind  /tmp/unwind/backtrace_test_libunwind_ptrace $(pidof dropbear) | head -n 20
FAILURE: unw_step() returned -10 for ip=10003c4c
pid: 21898
0000000076fe4ce4 __select                         offset 0x44 (sp=000000007fe47a50)
000000001000c8d4 _init                            offset 0x9964 (sp=000000007fe47aa0)
000000001001f024 _init                            offset 0x1c0b4 (sp=000000007fe47db0)
000000001001e240 _init                            offset 0x1b2d0 (sp=000000007fe47df0)
000000001001e2fc _init                            offset 0x1b38c (sp=000000007fe480f0)
0000000076f13e1c __libc_start_main                offset 0xfc (sp=000000007fe48120)
0000000010003c4c _init                            offset 0xcdc (sp=000000007fe48220)

$ LD_LIBRARY_PATH=/tmp/unwind  /tmp/unwind/backtrace_test_libunwind | head -n 20
#1  0x0000000010001218 sp=0x000000007ff1d7a0 fac(int) + 0x28
#2  0x0000000010001238 sp=0x000000007ff1d7c0 fac(int) + 0x48
#3  0x0000000010001238 sp=0x000000007ff1d7e0 fac(int) + 0x48
#4  0x0000000010001238 sp=0x000000007ff1d800 fac(int) + 0x48
#5  0x0000000010001238 sp=0x000000007ff1d820 fac(int) + 0x48
#6  0x0000000010001238 sp=0x000000007ff1d840 fac(int) + 0x48
#7  0x0000000010001238 sp=0x000000007ff1d860 fac(int) + 0x48
#8  0x0000000010001238 sp=0x000000007ff1d880 fac(int) + 0x48
#9  0x0000000010001238 sp=0x000000007ff1d8a0 fac(int) + 0x48
#10 0x0000000010001238 sp=0x000000007ff1d8c0 fac(int) + 0x48
#11 0x0000000010001238 sp=0x000000007ff1d8e0 fac(int) + 0x48
#12 0x00000000100013d0 sp=0x000000007ff1d900 main + 0x174
#13 0x0000000077317e1c sp=0x000000007ff1d990 __libc_start_main + 0xfc
#14 0x0000000010000f5c sp=0x000000007ff1da90 __start + 0x4c
#15 0x0000000010000f5c sp=0x000000007ff1da90 __start + 0x4c
#16 0x0000000010000f5c sp=0x000000007ff1da90 __start + 0x4c
#17 0x0000000010000f5c sp=0x000000007ff1da90 __start + 0x4c
#18 0x0000000010000f5c sp=0x000000007ff1da90 __start + 0x4c
#19 0x0000000010000f5c sp=0x000000007ff1da90 __start + 0x4c
#20 0x0000000010000f5c sp=0x000000007ff1da90 __start + 0x4c
```

As you can see, the remote unwinding with ptrace is basically correct, even though unw_step returned an error for the last entry. This error was not present in case 2, and may lead to something still incorrect in libunwind.

Note that the local unwinding case still suffers from the endless last entry.


So in conclusion, this commit fixes part of the problem but not all.

**Comment by ksergeyv** [comment]
@patrickdepinguin did You try to also revert #220 as I mentioned earlier?

**Comment by patrickdepinguin** [comment]
@ksergeyv  I am testing on top of libunwind 1.4.0, in that release the code of #220 (commit cee5505a99d42a423059a1add8b87b75aeb45daa) was not yet included.

I also need to clarify my earlier test results: I still had to specify: `ac_cv_have_decl_PTRACE_SETREGSET=no` to disable PTRACE_SETREGSET, as described in #173 .

**Comment by patrickdepinguin** [comment]
I now also tested current git master (a761757a31116729eb1d48fc0cdab23ce399dfc9) which includes the proposed commit already, still disabling PTRACE_SETREGSET, but get the same results as described above in case 3, with or without reverting commit cee5505a99d42a423059a1add8b87b75aeb45daa from #220 .

**Comment by ksergeyv** [comment]
@patrickdepinguin @kasperk81 it seems `unw_get_proc_info` should be
```
--- a/src/mips/Gget_proc_info.c
+++ b/src/mips/Gget_proc_info.c
@@ -32,11 +32,15 @@ unw_get_proc_info (unw_cursor_t *cursor, unw_proc_info_t *pi)
 
   ret = dwarf_make_proc_info (&c->dwarf);
   if (ret < 0) {
+#if _MIPS_SIM == _ABI64
     /* Construct a dummy proc info if Dwarf failed */
     memset (pi, 0, sizeof (*pi));
     pi->start_ip = c->dwarf.ip;
     pi->end_ip = c->dwarf.ip + 4;
     return 0;
+#else
+    return ret;
+#endif
```
to completely ignore changes from 5eec9a2ecb9a93996d566bbfbcdbe006f64b7e16 for O32 and N32.

**Comment by patrickdepinguin** [comment]
@ksergeyv  I applied the change you posted but this does not make a difference.
However, with that as hint, I looked further: in the patch you proposed in this PR and which is now on master, there is one difference compared to the original situation: 'return ret' vs. 'return 0' . 
If I apply below  patch on top of current master a761757a31116729eb1d48fc0cdab23ce399dfc9, then I get correct behavior as in case 2 in https://github.com/libunwind/libunwind/pull/275#issuecomment-877130861 .

```
diff --git a/src/mips/Gstep.c b/src/mips/Gstep.c
index 79fea71f..1a31f4fc 100644
--- a/src/mips/Gstep.c
+++ b/src/mips/Gstep.c
@@ -219,7 +219,7 @@ unw_step (unw_cursor_t *cursor)
 #if _MIPS_SIM == _ABI64
       return _step_n64(c);
 #else
-      return ret;
+      return 0;
 #endif
     }
 
```

**Comment by ksergeyv** [comment]
@patrickdepinguin I suppose your test should check `unw_step(cursor) <= 0` to stop unwinding according to `unw_step` description.

**Comment by kasperk81** [comment]
Based on the above looks like changing https://github.com/cslarsen/libunwind-examples/blob/f2905ae75e162f9254e1f818fdced6bf310baa19/backtrace.cpp#L21 to `while ( unw_step(&cursor) > 0 )` should fix #172.

@patrickdepinguin, it is also the case with other archs: https://github.com/libunwind/libunwind/blob/a761757a31116729eb1d48fc0cdab23ce399dfc9/src/x86_64/Gstep.c#L154-L159

**Comment by kasperk81** [comment]
> Based on the above looks like changing https://github.com/cslarsen/libunwind-examples/blob/f2905ae75e162f9254e1f818fdced6bf310baa19/backtrace.cpp#L21 to while ( unw_step(&cursor) > 0 ) should fix #172.

I meant applying  @ksergeyv's patch to libunwind and fixing `while`.

@ksergeyv, @patrickdepinguin, can you send a PR if it works?

ps, we should add that example in this repo if @cslarsen has no objection. I had to apply https://github.com/cslarsen/libunwind-examples/pull/1 to test that example on riscv64.

**Comment by patrickdepinguin** [comment]
From MIPS64n32 point of view: if we change the while in the test code, then the current master of libunwind is fine wrt issue #172. This master already includes the commit in this PR. The additional change mentioned by @ksergeyv  in comment https://github.com/libunwind/libunwind/pull/275#issuecomment-877152021  does not seem to make a difference in my test cases. I'm not familiar enough with the code to judge whether the 'dummy dwarf entry' introduced originally is really n64-specific, or also works fine in n32/o32 mode.

The example from cslarsen/libunwind-examples was really helpful for my testing, and having such example in the libunwind repo itself makes a lot of sense to me.

**Comment by ksergeyv** [comment]
@patrickdepinguin @kasperk81 I suppose only a test fix would be sufficient.
